File: installer/wix/generate-images.py
#!/usr/bin/env python3
"""
Generate placeholder BMP images for WiX installer.

This script creates the required banner.bmp (493x58) and dialog.bmp (374x316)
placeholder images with a simple branded appearance.

Requirements:
- Python 3.6+
- Pillow: pip install Pillow

Usage:
    python generate-images.py
"""

import struct
import sys


def create_bmp(width: int, height: int, filename: str, bg_color: tuple, text: str = "") -> None:
    """
    Create a simple BMP file with a solid background color.

    Args:
        width: Image width in pixels
        height: Image height in pixels
        filename: Output filename
        bg_color: Background color as (R, G, B) tuple
        text: Optional text (ignored in this simple implementation)
    """
    # BMP files store rows bottom-to-top, and each row is padded to 4-byte boundary
    row_size = (width * 3 + 3) & ~3  # 24-bit (3 bytes per pixel), padded to 4 bytes
    pixel_data_size = row_size * height
    file_size = 54 + pixel_data_size  # 54 byte header + pixel data

    with open(filename, 'wb') as f:
        # BMP Header (14 bytes)
        f.write(b'BM')                          # Signature
        f.write(struct.pack('<I', file_size))   # File size
        f.write(struct.pack('<HH', 0, 0))       # Reserved
        f.write(struct.pack('<I', 54))          # Pixel data offset

        # DIB Header (BITMAPINFOHEADER - 40 bytes)
        f.write(struct.pack('<I', 40))          # DIB header size
        f.write(struct.pack('<i', width))       # Width
        f.write(struct.pack('<i', height))      # Height (positive = bottom-up)
        f.write(struct.pack('<H', 1))           # Color planes
        f.write(struct.pack('<H', 24))          # Bits per pixel
        f.write(struct.pack('<I', 0))           # Compression (none)
        f.write(struct.pack('<I', pixel_data_size))  # Image size
        f.write(struct.pack('<i', 2835))        # X pixels per meter (72 DPI)
        f.write(struct.pack('<i', 2835))        # Y pixels per meter (72 DPI)
        f.write(struct.pack('<I', 0))           # Colors in color table
        f.write(struct.pack('<I', 0))           # Important colors

        # Pixel data (BGR format, bottom-to-top)
        r, g, b = bg_color
        row = bytes([b, g, r] * width)
        padding = bytes(row_size - width * 3)

        for _ in range(height):
            f.write(row + padding)

    print(f"Created {filename} ({width}x{height})")


def main():
    # Flight Hub brand colors
    PRIMARY_BLUE = (41, 98, 255)      # Vibrant blue
    DARK_BLUE = (26, 62, 161)         # Darker blue for dialog

    # Banner: 493x58 - shown at top of installer pages
    create_bmp(493, 58, "banner.bmp", PRIMARY_BLUE)

    # Dialog: 374x316 - shown on welcome/exit pages (left side)
    create_bmp(374, 316, "dialog.bmp", DARK_BLUE)

    print("\nPlaceholder images created successfully.")
    print("For a professional installer, replace these with proper branded images:")
    print("  - banner.bmp: 493x58 pixels, displayed at top of installer dialogs")
    print("  - dialog.bmp: 374x316 pixels, displayed on welcome/finish pages")


if __name__ == "__main__":
    main()

File: installer/wix/generate-images.rs
#!/usr/bin/env cargo +nightly -Zscript
// SPDX-License-Identifier: MIT OR Apache-2.0

//! Generate placeholder BMP images for the WiX installer.
//!
//! Usage:
//!   cargo +nightly -Zscript installer/wix/generate-images.rs
//!
//! The script writes `banner.bmp` (493x58) and `dialog.bmp` (374x316) in the
//! current directory, matching the historical generator behavior.

use std::fs::File;
use std::io::{self, Write};
use std::path::Path;

const PRIMARY_BLUE: Rgb = Rgb::new(41, 98, 255);
const DARK_BLUE: Rgb = Rgb::new(26, 62, 161);

#[derive(Clone, Copy)]
struct Rgb {
    red: u8,
    green: u8,
    blue: u8,
}

impl Rgb {
    const fn new(red: u8, green: u8, blue: u8) -> Self {
        Self { red, green, blue }
    }
}

fn main() -> io::Result<()> {
    create_bmp(493, 58, "banner.bmp", PRIMARY_BLUE)?;
    create_bmp(374, 316, "dialog.bmp", DARK_BLUE)?;

    println!("\nPlaceholder images created successfully.");
    println!("For a professional installer, replace these with proper branded images:");
    println!("  - banner.bmp: 493x58 pixels, displayed at top of installer dialogs");
    println!("  - dialog.bmp: 374x316 pixels, displayed on welcome/finish pages");

    Ok(())
}

fn create_bmp(width: u32, height: u32, filename: &str, bg_color: Rgb) -> io::Result<()> {
    let row_size = (width * 3 + 3) & !3;
    let pixel_data_size = row_size * height;
    let file_size = 54 + pixel_data_size;

    let mut file = File::create(Path::new(filename))?;

    file.write_all(b"BM")?;
    file.write_all(&file_size.to_le_bytes())?;
    file.write_all(&0u16.to_le_bytes())?;
    file.write_all(&0u16.to_le_bytes())?;
    file.write_all(&54u32.to_le_bytes())?;

    file.write_all(&40u32.to_le_bytes())?;
    file.write_all(&(width as i32).to_le_bytes())?;
    file.write_all(&(height as i32).to_le_bytes())?;
    file.write_all(&1u16.to_le_bytes())?;
    file.write_all(&24u16.to_le_bytes())?;
    file.write_all(&0u32.to_le_bytes())?;
    file.write_all(&pixel_data_size.to_le_bytes())?;
    file.write_all(&2835i32.to_le_bytes())?;
    file.write_all(&2835i32.to_le_bytes())?;
    file.write_all(&0u32.to_le_bytes())?;
    file.write_all(&0u32.to_le_bytes())?;

    let mut row = Vec::with_capacity(row_size as usize);
    for _ in 0..width {
        row.extend_from_slice(&[bg_color.blue, bg_color.green, bg_color.red]);
    }
    row.resize(row_size as usize, 0);

    for _ in 0..height {
        file.write_all(&row)?;
    }

    println!("Created {filename} ({width}x{height})");
    Ok(())
}
